baseplate/frameworks/queue_consumer/kombu.py

```diff
@@ -124,22 +162,110 @@ def __init__(
         name: str,
         handler_fn: Handler,
         error_handler_fn: Optional[ErrorHandler] = None,
+        retry_mode: RetryMode = RetryMode.REQUEUE,
+        retry_limit: Optional[int] = None,
     ):
         self.baseplate = baseplate
         self.name = name
         self.handler_fn = handler_fn
         self.error_handler_fn = error_handler_fn
+        self.retry_mode = retry_mode
+        self.retry_limit = retry_limit
+
+    def _is_error_recoverable(self, exc: Exception) -> bool:
+        if isinstance(exc, KnownException):
+            return exc.is_recoverable()
+        return True  # for backward compatibility, retry unexpected errors
+
+    def _handle_error(
+        self,
+        message: kombu.Message,
+        prometheus_labels: AmqpConsumerPrometheusLabels,
+        exc: Exception,
+    ) -> None:
+        if not self._is_error_recoverable(exc):
+            message.reject()
+            logger.exception(
+                "Unrecoverable error while trying to process a message.  The message has been discarded."
+            )
+            return
+
+        message_exchange = message.delivery_info.get("exchange")
+        message_routing_key = message.delivery_info.get("routing_key")
+        if not message_exchange or not message_routing_key or self.retry_mode == RetryMode.REQUEUE:
+            logger.exception(
+                "Recoverable error while trying to process a message. "
+                "The message has been returned to the queue broker."
+            )
+            message.requeue()
+            return
```

Comment:
Having the queue always re-enqueue failed messages opens the system up to infinite retries. Rather than retrying failed messages infinitely we usually suggest using dead-letter queues. I would consider infinite retries to be actively harmful to the system.

Reply:
It's a pre-existing behavior which makes my changes backward compatible; in current version it looks like
```
            if self.error_handler_fn:
                self.error_handler_fn(context, message_body, message, exc)
            else:
                message.requeue()
```